pkg/download/bucket/bucket.go

```diff
@@ -0,0 +1,157 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bucket
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api/clients/buckets"
+	tools "github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/buckettools"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/idutils"
+	jsonutils "github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/json"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log/field"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/raw"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/coordinate"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/parameter"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/template"
+	v2 "github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/project/v2"
+)
+
+type BucketClient interface {
+	List(ctx context.Context) (buckets.ListResponse, error)
+}
+
+type Downloader struct {
+	client BucketClient
+}
+
+func NewDownloader(client BucketClient) *Downloader {
+	return &Downloader{
+		client: client,
+	}
+}
+func (d *Downloader) Download(projectName string, _ ...config.BucketType) (v2.ConfigsPerType, error) { // error in return is just to complain to interface
+	result := make(v2.ConfigsPerType)
+	response, err := d.client.List(context.TODO())
+	if err != nil {
+		log.WithFields(field.Type("bucket"), field.Error(err)).Error("Failed to fetch all bucket definitions: %v", err)
+		return nil, nil
+	}
+
+	if apiErr, isErr := response.AsAPIError(); isErr {
+		log.WithFields(field.Type("bucket"), field.Error(apiErr)).Error("Failed to fetch all bucket definitions: %v", apiErr)
+		return nil, nil
+	}
+
+	configs := d.convertAllObjects(projectName, response.Objects)
+	result["bucket"] = configs
+	return result, nil
+}
+
+func (d *Downloader) convertAllObjects(projectName string, objects [][]byte) []config.Config {
+	result := make([]config.Config, 0, len(objects))
+	for _, o := range objects {
+
+		c, err := convertObject(o, projectName)
+		if err != nil {
+			log.WithFields(field.Coordinate(coordinate.Coordinate{Project: c.Coordinate.Project, Type: "bucket", ConfigId: c.Coordinate.ConfigId}), field.Error(err)).
+				Warn("Failed to get configuration for %v (%s): %v", c.Coordinate.ConfigId, "bucket", err)
+			continue
+		}
+
+		if c == nil {
+			continue
```

Comment:
IMO it would be worth a debug log here that a default value is being skipped.

Thinking about single-responsibility, I wouldn't actually expect `convertObject` to handle skipping default values. 
Just having convert return the config and changing this check from `c == nil` to 
```go
	// exclude builtin bucket names
		if tools.IsDefault(c.OriginObjectId) {
			log.Debug("Skipping download of immutable default Bucket %s", c.OriginObjectId)
			continue
		}
```
would IMO be nicer separation of concerns and make it easier to follow what is happening here. 
To understand what a nil return actually means and why it's skipped, I currently need to read convertObject.